NumLib/ODESolver/NonlinearSolver.cpp

```diff
@@ -184,6 +184,17 @@ void NonlinearSolver<NonlinearSolverTag::Newton>::assemble(
     //      equation every time and could not forget it.
 }
 
+void NonlinearSolver<NonlinearSolverTag::Newton>::calculateOutOfBalanceForces(
```

Comment:
IMO, NumLib should be independent of physics.

Reply:
I just updated the documentation, there are some more details.
In the document we named the initial residuum F_oob -- out-of-balance forces. These are not the "physical" forces, but general external loads. For mechanics, these are indeed forces as
```tex
$F^oob = r_0 = \int\Omega B^T \sigma_0\; d\Omega$
```

Reply:
Thanks for the doc. Is there any way to see the documentation without doing git lfs? 

Do we need this calculation besides mechanics?  To me, "load", "force" sounds specific to mechanics.

Reply:
> the doc.

https://deploy-preview-2633--ogs.netlify.com/docs/benchmarks/small-deformations/mechanics-linear-nonequilibrium-states/

The preview to the benchmark sites is under the "deploy/netlify -- Details" in the checks of the PR. There in "docs", "Selected Benchmarks", you will find the above site.

> "load", "force" ...

Same arguments are valid for pressure or saturation. One could rename Foob to `initial_residuum`, `r_0`, or something similar to avoid mechanics relation...

Reply:
@endJunction The F^oob equation above also contains body forces and Neumann bc's right? I think you were just brief but want to make sure...

@norihiro-w I sent you the equations a while ago to replace the non-equ. stresses with a better more general solution. Indeed "forces" points to mechanics, like much of the FE nomenclature (e.g. stiffness matrix). So I wouldn't have a problem with that name. The most general alternative could be "out of balance fluxes". That applies to all processes. In case of mechanics, forces are linear momentum fluxes.... then we could even keep the F ;-)

Reply:
and yes, it is intended that this applies to all processes where needed ... we can write this for a general balance equation and Foob would always be the initial out of balance flux that a user may want to drive an initial process or not (and hence would have to compensate) ... 
time for dinner now in China ...

Reply:
@endJunction thank you for the tip. I could read it. 

@endJunction @nagelt Indeed I'm not suggesting renaming the function. I'm suggesting moving this functionality to somewhere in ProcessLib. NumLib should be a general pure numerical library and we should avoid contaminate it with physics except a concept of space and time. I guess you can use a post-assembly function caller in the nonlinear solver class without changing the general interface. If you still want to change NumLib, it would be better if you use a concept in math/numeric, such as residual shift/offset (i'm not sure this example is appropriate). 

@nagelt Is "Foob" a word you defined? I was asking Dima about the name because I'm surprised that this trivial algorithm has a name.

Reply:
@norihiro-w maybe I misunderstand you and say something totally obvious to you now, but we just use "Foob" as a short-hand for $F_oob$ = out-of-balance-forces/fluxes. We could also call this initial residual if we leave physics out of the name ... but for most engineers the physical name implies the concept better than the purely mathematical one..

regarding ProcessLib: if we can move it there without much extra code into a general place, I agree. I'm not fully aware of the structure there and the effort involved, so I'll leave this call to @endJunction